db/migrate/20211026145034_create_users.rb

```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: false
+
+# Migration to create a users table
+class CreateUsers < ActiveRecord::Migration[6.1]
+  def change
+    create_table :users do |t|
+      t.string :username
+      t.string :password
```

Comment:
Storing passwords like this is a really bad idea. In this case, you're storing the password in plain text. What that means is that if somebody breaks in to the database, they can steal all of the passwords.

When this happens, the password lists often get posted on the Internet and become common knowledge so anybody can log in and pretend to be that user.

What makes this worse is that a lot of people use the same username and password on lots of systems. That means that it is likely the compromise will actually effect other people's websites too...

What is generally considered to be "good practice" now is to use a non-reversible hash function and store the hashed value in the database instead. This means that instead of storing `'pawsome'` we end up storing something like `hash('pawsome')`. Then when we want to check the password matches, we could do something like `hash(params[:password]) == User.find(params[:username]).password`. Because the hash function is one-way, if our database is compromised, the attacker can't actually get the password.

This is a massive over-simplification and still not great as-is. Imagine that you have 2 users in your system who both have the same password.... If we store it like this, then the attacker now knows that if they can guess one password, they can guess both. Because hash functions are standardised and commonplace, it would be likely that somebody out there would have used that password and it would appear in a list somewhere with the hashed version and plain text, so the attacker would know what the password is.... To get around this, we would use `salt` and `pepper` (https://en.wikipedia.org/wiki/Salt_(cryptography)) to make each hashed value unique.

Fortunately, there are a lot of libraries that help us get around some of these issues. In Rails, we typically use `devise` which is a gem that does most of the authentication stuff for us :)